### PR TITLE
Support a generic --run-on flag

### LINKED ISSUE #6185: Accept a `--run-on` flag
/kind feature

## Which functionality do you think we should add?

`odo` should accept a `run-on` flag generically, in the same way it accepts a `-o` flag.

The value of the flag can be either `cluster` or `podman`

The value of the flag should be passed in the `ctx` received by Run (and make Complete and Validate receive this `ctx` too).

## Why is this needed?

The generic code (`genericlioptions.GenericRun`) needs to know the targeted platform to inject the correct package (podman of kubernetes) during dependency injection.

Part of #6151

### PR DESCRIPTION
**Pull request by feloy**
**What type of PR is this:**

/kind feature

**What does this PR do / why we need it:**

- adds a new generic option `--run-on`.
- ctx is passed to Complete and Validate methods
- the presence of the -o json flag is saved in the ctx passed to Complete, Validate and Run
- the value of the --run-on flag is saved in the ctx passed to Complete, Validate and Run

**Which issue(s) this PR fixes:**

Fixes #6185 

**PR acceptance criteria:**

- [ ] Unit test 

- [ ] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer:**

### CODE REVIEW

File: pkg/odo/commonflags/run_on.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+func CheckRunOnCommand(cmd *cobra.Command) error {
+
+	// Get the needed values
+	runOnFlag := pflag.Lookup(RunOnFlagName)
+	hasFlagChanged := runOnFlag != nil && runOnFlag.Changed
+	runOn := cmd.Annotations["runOn"]
+
+	// Check the valid output
+	if hasFlagChanged && runOnFlag.Value.String() != RunOnPodman && runOnFlag.Value.String() != RunOnCluster {
+		return fmt.Errorf("%s is not a valid target platform for --run-on, please select either cluster (default) or podman", runOnFlag.Value.String())
```
Comment (rm3l) [inline pkg/odo/commonflags/run_on.go:48]: Just to make it clear what the expected values are.
Also, even though #6150 will be adding an experimental mode, meanwhile, I think we could make it clear that `podman` support is experimental:

```suggestion
		return fmt.Errorf("%s is not a valid target platform for --run-on, please select either 'cluster' (default) or 'podman' (experimental)", runOnFlag.Value.String())
```

File: pkg/odo/commonflags/run_on.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+	cmd.Annotations["runOn"] = "true"
+}
+
+// AddRunOnFlag adds the --run-on output flag to all commands
+// We use "flag" in order to make this accessible throughtout ALL of odo, rather than the
+// above traditional "persistentflags" usage that does not make it a pointer within the 'pflag'
+// package
+func AddRunOnFlag() {
+	flag.CommandLine.String(RunOnFlagName, "", "Specify target platform, supported platforms: cluster(default), podman")
```
Comment (rm3l) [inline pkg/odo/commonflags/run_on.go:34]: Same suggestion as above:

```suggestion
	flag.CommandLine.String(RunOnFlagName, "", "Specify target platform, supported platforms: 'cluster' (default), 'podman' (experimental)")
```

File: pkg/odo/cli/init/init.go [hunk 3]
```diff
@@ -113,7 +114,7 @@ func (o *InitOptions) Validate() error {
 		return err
 	}
 
-	if len(o.flags) == 0 && log.IsJSON() {
+	if len(o.flags) == 0 && fcontext.GetOutput(ctx) {
```
Comment (rm3l) [inline pkg/odo/cli/init/init.go:117]: Would it make sense to have this function named `fcontext.IsJsonOutput`? Or make `GetOutput` return the actual output value as a constant?
(I am thinking of cases where we might need to support other output types in the future.)

File: pkg/odo/commonflags/context/context.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	runOnKey  runOnKeyType
+)
+
+// WithOutput sets the value for the output flag (-o) in ctx
+func WithOutput(ctx context.Context, val bool) context.Context {
+	return context.WithValue(ctx, outputKey, val)
+}
+
+// GetOutput gets value of output flag (-o) in ctx
+func GetOutput(ctx context.Context) bool {
```
Comment (rm3l) [inline pkg/odo/commonflags/context/context.go:29]: Same comment as above:

Would it make sense to have this function named `IsJsonOutput`? Or make it return the actual output value as a constant?
(I am thinking of cases where we might need to support other output types in the future.)

File: pkg/odo/commonflags/context/context.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	runOnKeyType  struct{}
+)
+
+var (
+	outputKey outputKeyType
+	runOnKey  runOnKeyType
+)
+
+// WithOutput sets the value for the output flag (-o) in ctx
+func WithOutput(ctx context.Context, val bool) context.Context {
```
Comment (rm3l) [inline pkg/odo/commonflags/context/context.go:24]: Same question: to make the intent clearer, would it make sense to rename this into something like `WithJsonOutput`?

File: pkg/odo/commonflags/output.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		v := flag.CommandLine.Lookup("v").Value.String()
+		if level, ok := os.LookupEnv("ODO_LOG_LEVEL"); ok && v == "0" {
+			_ = flag.CommandLine.Set("v", level)
+		}
+	}
+	return nil
+}
+
+// GetOutputValue returns true if -o flag is used and value is "json"
+func GetOutputValue(cmd cmdline.Cmdline) bool {
```
Comment (rm3l) [inline pkg/odo/commonflags/output.go:82]: Same question: to make the intent clearer, would it make sense to rename this into something like `GetJsonOutputValue`?

**Review by rm3l — CHANGES_REQUESTED** [review]

File: pkg/odo/commonflags/output.go (hunk 6 above)
Comment (feloy) [reply pkg/odo/commonflags/output.go:82]: I was not really happy with these function names, thanks for the suggestion

**Review by rm3l — APPROVED** [review]

**Pull request merged**